File: leetcode/[1319] Number of Operations to Make Network Connected/revision.js
/*
 * @lc app=leetcode id=1319 lang=javascript
 *
 * [1319] Number of Operations to Make Network Connected
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var makeConnected = function(n, connections) {
    const disjointSet = Array(n).fill(-1);
    let numOfCycles = 0;
    connections.forEach(([a, b]) => {
        if (_formingCycle(disjointSet, a, b)) {
            numOfCycles++;
        } else {
            _union(disjointSet, a, b);
        }
    });
    const numOfUnconnected = disjointSet.filter(v => v < 0).length - 1;
    return numOfUnconnected ? (numOfCycles >= numOfUnconnected ? numOfUnconnected : -1) : 0;
};

const _findRoot = (set, a) => {
    if (set[a] < 0) return a;
    const root = _findRoot(set, set[a]);
    set[a] = root; //path compression
    return root;
};

const _formingCycle = (set, a, b) => {
    return _findRoot(set, a) === _findRoot(set, b);
}

const _union = (set, a, b) => {
    const aRoot = _findRoot(set, a);
    const bRoot = _findRoot(set, b);
    if (aRoot !== bRoot) {
        set[aRoot] = bRoot;
    }
};

// @lc code=end

File: leetcode/[200] Number of Islands/revision.java
/*
 * @lc app=leetcode id=200 lang=java
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int ans = 0;
        for (int x = 0; x < grid.length; x++)
            for (int y = 0; y < grid[0].length; y++) {
                if (!visited[x][y] && grid[x][y] == '1')
                    ans++;
                DFS(visited, grid, x, y);
            }
        return ans;
    }

    private void DFS(boolean[][] visited, char[][] grid, int x, int y) {
        if (x < 0 || y < 0 || x >= grid.length || y >= grid[0].length || visited[x][y] || grid[x][y] == '0')
            return;
        visited[x][y] = true;
        DFS(visited, grid, x + 1, y);
        DFS(visited, grid, x - 1, y);
        DFS(visited, grid, x, y + 1);
        DFS(visited, grid, x, y - 1);
    }
}
// @lc code=end
